crates/diesel_models/src/query/role.rs

```diff
@@ -104,32 +93,72 @@ impl Role {
         .await
     }
 
-    pub async fn list_roles(
+    pub async fn find_role_by_role_name_in_lineage(
         conn: &PgPooledConn,
-        merchant_id: &id_type::MerchantId,
-        org_id: &id_type::OrganizationId,
-    ) -> StorageResult<Vec<Self>> {
-        let predicate = dsl::org_id.eq(org_id.to_owned()).and(
-            dsl::scope.eq(RoleScope::Organization).or(dsl::merchant_id
-                .eq(merchant_id.to_owned())
-                .and(dsl::scope.eq(RoleScope::Merchant))),
-        );
+        role_name: &str,
+        entity_type: ListRolesByEntityPayload,
+    ) -> StorageResult<Self> {
+        let mut query = <Self as HasTable>::table().into_boxed();
 
-        generics::generic_filter::<<Self as HasTable>::Table, _, _, _>(
-            conn,
-            predicate,
-            None,
-            None,
-            Some(dsl::last_modified_at.asc()),
+        if let Some(org_id) = entity_type.get_organization_id() {
+            query = query.filter(
+                dsl::org_id
+                    .eq(org_id)
+                    .and(dsl::scope.eq(RoleScope::Organization))
+                    .and(dsl::entity_type.eq_any(vec![
+                        EntityType::Profile,
+                        EntityType::Merchant,
+                        EntityType::Organization,
```

Comment:
These entity types should be based on the Payload's entity type.

If we run this query for profile level Payload, we will also search for org level entity_types which is wrong (correct me if I am wrong).

And if it is not based on the payload, then this query is technically a subset for the list query.

Reply:
Its been modified to check with list query